README.md
# JWT Encrypt

A JWT encryption and decryption library for node applications

This library provides a more granular encryption as it strives solve the vulnarabilities in the crypto encryption and descryption by using the latest `createCipheriv` and `createDecipheriv` as per the deprecation notice on `createCipher` and `createDecipher` which is currently semantically insecure for all supported ciphers and fattally flawed for ciphers in counter mode such as `(CTR, GCM or CCM)`.

## Installation

```bash
npm install jwt-encrypt
```

or

```bash
yarn add jwt-encrypt
```

## Encryption algorithms supported

The encryption algorithms determine the key. The following shows the list algorithms supported and their key and iv(initialization vector) lengths required

| **Algorithm** | **Key Length** | **IV Length** |
| ------------- | -------------- | ------------- |
| aes-256-cbc   | 32             | 16            |
| aes-192-cbc   | 24             | 16            |
| aes-128-cbc   | 16             | 16            |

### Encryption options

```ts
{
  key: string; // a 16-bit, 24-bit, or 32-bit raw string used by the algorithms
  iv: string; // a 16-bit raw string initialization vector
  algorithm: 'aes-128-cbc' | 'aes-192-cbc' | 'aes-256-cbc'; // The cypher algorithm to be used to  encrypt the payload
}
```

## Usage

### Sign Token

(Synchronous) Returns the JsonWebToken containing encrypted payload as string

```ts
import jwte from 'jwt-encrypt';

const token = jwte.sign(payload, jwtSecret, encryptionOptions, jwtOptions);
```

#### Example Usage

```ts
import jwte from 'jwt-encrypt';

const payload = { test: 'test' };
const jwtSecret = 'secret';
const encryptionOptions = {
  algorithm: 'aes-256-cbc',
  iv: 'abcd1234abcd1234', // 16-bit string
  key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
};

const token = jwte.sign(payload, jwtSecret, encryptionOptions, {
  expiresIn: '1min',
});
```

### Decode Token

(Synchronously) Decode a given token with encrypted data.

_NB: While decoding you must provide the `algorithm`, `key` and `iv` that were used while signing the token_

```ts
import jwte from 'jwt-encrypt';

const decoded = jwte.decode(token, encryptionOptions);
```

#### Example Usage

```ts
import jwte from 'jwt-encrypt';

const payload = { test: 'test' };

const encryptionOptions = {
  algorithm: 'aes-256-cbc',
  iv: 'abcd1234abcd1234', // 16-bit string
  key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
};

const decoded = jwte.decode(token, encryptionOptions);
```

### Verify Token

(Asynchronous) If a callback is supplied, function acts asynchronously. The callback is called with the decoded payload if the signature is valid and optional expiration, audience, or issuer are valid. If not, it will be called with the error.

_NB: While verifying you must provide the `algorithm`, `key` and `iv` that were used while signing the token_

```ts
import jwte from 'jwt-encrypt';

jwte.verify(signedToken, jwtSecret, encryptionOptions, jwtOptions, (err, verifiedToken) => {});
```

#### Example Usage

```ts
import jwte from 'jwt-encrypt';

const payload = { test: 'test' };
const jwtSecret = 'secret';

const encryptionOptions = {
  algorithm: 'aes-256-cbc',
  iv: 'abcd1234abcd1234', // 16-bit string
  key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
};

const jwtOptions = {};

jwte.verify(signedToken, jwtSecret, encryptionOptions, jwtOptions, (err, verifiedToken) => {
  if (err) {
    console.log('Error', err);
  }

  console.log('VerifiedToken', verifiedToken);
});
```

## License

MIT


CHANGELOG.md
# CHANGELOG

## 1.0.3 (Mar 2, 2023)

- Bug fixes and testing
- Initial public release


__tests__/util.test.ts
import { Cipher } from '../utils';

import type { EncryptionOptions } from '../types';

describe('"Cipher"', () => {
  describe('"Cipher.encrypt"', () => {
    it('should encrypt a given payload', () => {
      const payload = 'Test';
      const options: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };
      const expectedEncryptedResponse = 'ef26b4ac0b833be8a925b005a2e5ad45';

      const encryptedPayload = Cipher.encrypt(payload, options);

      expect(encryptedPayload).toBeDefined();
      expect(encryptedPayload).toHaveProperty('data');
      expect(typeof encryptedPayload.data).toBe('string');
      expect(encryptedPayload.data).toBe(expectedEncryptedResponse);
    });

    it('should throw encryption error when encrypting a given payload with wrong algorithm', () => {
      const payload = 'Test';
      const options: EncryptionOptions = {
        // Deliberately force the algorithm to a different algorithm to catch encryption errors
        //@ts-expect-error (2820): FIXME: Type '"aes-256-ccm"' is not assignable to type 'EncryptionAlgorithm'. Did you mean '"aes-256-cbc"
        algorithm: 'aes-256-ccm',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const errorMessage = 'Error encrypting string. Please try again';

      expect(() => {
        Cipher.encrypt(payload, options);
      }).toThrowError(errorMessage);
    });
  });

  describe('"Cipher.decrypt"', () => {
    it('should decrypt a given encrypted payload', () => {
      const payload = 'Test';
      const options: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };
      const expectedEncryptedResponse = 'ef26b4ac0b833be8a925b005a2e5ad45';

      const encryptedPayload = Cipher.encrypt(payload, options);

      expect(encryptedPayload).toBeDefined();
      expect(encryptedPayload).toHaveProperty('data');
      expect(typeof encryptedPayload.data).toBe('string');
      expect(encryptedPayload.data).toBe(expectedEncryptedResponse);

      const decryptedPayload = Cipher.decrypt(encryptedPayload.data, options);

      expect(decryptedPayload).toBe(payload);
    });

    it('should throw an error when decrypting a given encrypted payload with wrong decryption options', () => {
      const payload = 'Test';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const decryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'bbcd1234abcd1235', // 16-bit string
        key: 'bbcd1234abcd1234efgh5678efgh5670', // 32-bit string
      };

      const expectedEncryptedResponse = 'ef26b4ac0b833be8a925b005a2e5ad45';
      const errorMessage =
        'Error decrypting string. Ensure the text parameter is an encrypted string or the key and iv (initialization vector) are the same that were provided while encrypting';

      const encryptedPayload = Cipher.encrypt(payload, encryptionOptions);

      expect(encryptedPayload).toBeDefined();
      expect(encryptedPayload).toHaveProperty('data');
      expect(typeof encryptedPayload.data).toBe('string');
      expect(encryptedPayload.data).toBe(expectedEncryptedResponse);

      expect(() => {
        Cipher.decrypt(encryptedPayload.data, decryptionOptions);
      }).toThrowError(errorMessage);
    });
  });
});


__tests__/jwtEncrypt.test.ts
import jwte from '../index';

import type { EncryptionOptions } from '../types';

describe('"sign" - signing a token asynchronously', () => {
  describe('"sign" jwt token success', () => {
    it('should to return a signed jwt token when using "aes-256-cbc" encryption algorithm', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');
    });

    it('should to return a signed jwt token when using "aes-192-cbc" encryption algorithm', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-192-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678', // 24-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');
    });

    it('should to return a signed jwt token when using "aes-128-cbc" encryption algorithm', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-128-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234', // 16-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');
    });
  });

  describe('"sign" jwt token error on wrong key', () => {
    it('should throw an error signing a jwt token using "aes-256-cbc" encryption algorithm wihout providing a key', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: '',
      };

      const errorMessage = 'Missing encyption key';

      expect(() => {
        jwte.sign(payload, jwtSecret, encryptionOptions, {
          expiresIn: '1min',
        });
      }).toThrowError(errorMessage);
    });

    it('should throw an error signing a jwt token using "aes-256-cbc" encryption algorithm with a key not equal to 32-bit string', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh567', // 31-bit string
      };

      const errorMessage = 'Encryption key must be a 32-bit string. Received 31-bit string';

      expect(() => {
        jwte.sign(payload, jwtSecret, encryptionOptions, {
          expiresIn: '1min',
        });
      }).toThrowError(errorMessage);
    });

    it('should throw an error signing a jwt token using "aes-192-cbc" encryption algorithm with a key not equal to 24-bit string', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-192-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678e', // 25-bit string
      };

      const errorMessage = 'Encryption key must be a 24-bit string. Received 25-bit string';

      expect(() => {
        jwte.sign(payload, jwtSecret, encryptionOptions, {
          expiresIn: '1min',
        });
      }).toThrowError(errorMessage);
    });

    it('should throw an error signing a jwt token using "aes-128-cbc" encryption algorithm with a key not equal to 16-bit string', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-128-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd123', // 15-bit string
      };

      const errorMessage = 'Encryption key must be a 16-bit string. Received 15-bit string';

      expect(() => {
        jwte.sign(payload, jwtSecret, encryptionOptions, {
          expiresIn: '1min',
        });
      }).toThrowError(errorMessage);
    });
  });

  describe('"sign" jwt token error on wrong iv (initialization vector)', () => {
    it('should throw an error signing a jwt token using "aes-256-cbc" encryption algorithm without providing an iv', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: '',
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const errorMessage = 'Missing encryption iv(initialization vector)';

      expect(() => {
        jwte.sign(payload, jwtSecret, encryptionOptions, {
          expiresIn: '1min',
        });
      }).toThrowError(errorMessage);
    });

    it('should throw an error signing a jwt token using "aes-256-cbc" encryption algorithm with an iv not equal to 16-bit string', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234a', // 15-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const errorMessage = 'Encryption iv(initialization vector) must be a 16-bit string: Received 17-bit string';

      expect(() => {
        jwte.sign(payload, jwtSecret, encryptionOptions, {
          expiresIn: '1min',
        });
      }).toThrowError(errorMessage);
    });

    it('should throw an error signing a jwt token using "aes-192-cbc" encryption algorithm with an iv not equal to 16-bit string', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-192-cbc',
        iv: 'abcd1234abcd123', // 15-bit string
        key: 'abcd1234abcd1234efgh5678', // 24-bit string
      };

      const errorMessage = 'Encryption iv(initialization vector) must be a 16-bit string: Received 15-bit string';

      expect(() => {
        jwte.sign(payload, jwtSecret, encryptionOptions, {
          expiresIn: '1min',
        });
      }).toThrowError(errorMessage);
    });

    it('should throw an error signing a jwt token using "aes-128-cbc" encryption algorithm with an iv not equal to 16-bit string', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';
      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-128-cbc',
        iv: 'abcd1234abcd1234abcd', // 20-bit string
        key: 'abcd1234abcd1234', // 16-bit string
      };

      const errorMessage = 'Encryption iv(initialization vector) must be a 16-bit string: Received 20-bit string';

      expect(() => {
        jwte.sign(payload, jwtSecret, encryptionOptions, {
          expiresIn: '1min',
        });
      }).toThrowError(errorMessage);
    });
  });
});

describe('"decode" - decoding a signed jwt token without verifying if its valid', () => {
  describe('"decode" signed jwt token success', () => {
    it('should to return decoded value when using "aes-256-cbc" encryption algorithm', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      const decodedToken = jwte.decode(signedToken, encryptionOptions);

      expect(decodedToken).toHaveProperty('test');
      expect(decodedToken).toHaveProperty('iat');
      expect(decodedToken).toHaveProperty('exp');
      expect(typeof decodedToken).not.toBe('string');
      expect(typeof decodedToken !== 'string' && decodedToken?.test).toStrictEqual(payload.test);
    });

    it('should to return decoded value when using "aes-192-cbc" encryption algorithm', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-192-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678', // 24-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      const decodedToken = jwte.decode(signedToken, encryptionOptions);

      expect(decodedToken).toHaveProperty('test');
      expect(decodedToken).toHaveProperty('iat');
      expect(decodedToken).toHaveProperty('exp');
      expect(typeof decodedToken).not.toBe('string');
      expect(typeof decodedToken !== 'string' && decodedToken?.test).toStrictEqual(payload.test);
    });

    it('should to return decoded value when using "aes-128-cbc" encryption algorithm', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-128-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234', // 16-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      const decodedToken = jwte.decode(signedToken, encryptionOptions);

      expect(decodedToken).toHaveProperty('test');
      expect(decodedToken).toHaveProperty('iat');
      expect(decodedToken).toHaveProperty('exp');
      expect(typeof decodedToken).not.toBe('string');
      expect(typeof decodedToken !== 'string' && decodedToken?.test).toStrictEqual(payload.test);
    });

    it('should to return complete decoded value when jwt option complete set as true ', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      const decodedToken = jwte.decode(signedToken, encryptionOptions, {
        complete: true,
      });

      expect(typeof decodedToken).not.toBe('string');

      if (typeof decodedToken !== 'string') {
        expect(decodedToken).toHaveProperty('payload');
        expect(decodedToken).toHaveProperty('header');
        expect(decodedToken).toHaveProperty('signature');
        expect(decodedToken?.payload).toHaveProperty('test');
        expect(decodedToken?.header).toHaveProperty('alg');
        expect(decodedToken?.header).toHaveProperty('typ');

        expect(decodedToken?.payload?.test).toStrictEqual(payload.test);
        expect(decodedToken?.header?.alg).toEqual('HS256');
        expect(decodedToken?.header?.typ).toEqual('JWT');
      }
    });

    it('should to return a string decoded value ', () => {
      const payload = 'test';
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      const decodedToken = jwte.decode(signedToken, encryptionOptions);

      expect(typeof decodedToken).toBe('string');
      expect(decodedToken).toBe(payload);
      expect(typeof decodedToken).not.toBe('object');
    });

    it('should to return a string decoded payload value when jwt option complete is set as true ', () => {
      const payload = 'test';
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      const decodedToken = jwte.decode(signedToken, encryptionOptions, {
        complete: true,
      });

      expect(typeof decodedToken).not.toBe('string');
      expect(typeof decodedToken).toBe('object');

      if (typeof decodedToken !== 'string') {
        expect(decodedToken).toHaveProperty('payload');
        expect(decodedToken).toHaveProperty('header');
        expect(decodedToken).toHaveProperty('signature');
        expect(typeof decodedToken?.payload).toBe('string');
        expect(decodedToken?.header).toHaveProperty('alg');
        expect(decodedToken?.header).toHaveProperty('typ');

        expect(decodedToken?.payload).toStrictEqual(payload);
        expect(decodedToken?.header?.alg).toEqual('HS256');
        expect(decodedToken?.header?.typ).toEqual('JWT');
      }
    });
  });
});

describe('"verify" - decoding a signed jwt token and verifying if its valid', () => {
  describe('"verify" signed jwt token success', () => {
    it('should to return a verified signed token value', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      jwte.verify(signedToken, jwtSecret, encryptionOptions, {}, (err, verifiedToken) => {
        expect(verifiedToken).toHaveProperty('test');
        expect(verifiedToken).toHaveProperty('iat');
        expect(verifiedToken).toHaveProperty('exp');
        expect(typeof verifiedToken).not.toBe('string');
        expect(
          ///@ts-expect-error (2339) FIXME: Property 'test' does not exist on type 'string | JwtPayload | Jwt'.
          verifiedToken?.test,
        ).toStrictEqual(payload.test);
      });
    });

    it('should to return a complete verified signed token value when jwt option complete is set as true', () => {
      const payload = { test: 'test' };
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      jwte.verify(signedToken, jwtSecret, encryptionOptions, { complete: true }, (err, verifiedToken) => {
        expect(err).toBeNull();
        expect(typeof verifiedToken).not.toBe('string');

        if (typeof verifiedToken !== 'string') {
          expect(verifiedToken).toHaveProperty('payload');
          expect(verifiedToken).toHaveProperty('header');
          expect(verifiedToken).toHaveProperty('signature');
          expect(verifiedToken?.payload).toHaveProperty('test');
          expect(verifiedToken?.header).toHaveProperty('alg');
          expect(verifiedToken?.header).toHaveProperty('typ');

          expect(verifiedToken?.payload?.test).toStrictEqual(payload.test);
          expect(verifiedToken?.header?.alg).toEqual('HS256');
          expect(verifiedToken?.header?.typ).toEqual('JWT');
        }
      });
    });

    it('should return a verified signed token value when the payload is set as a string', () => {
      const payload = 'test';
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      jwte.verify(signedToken, jwtSecret, encryptionOptions, {}, (err, verifiedToken) => {
        expect(err).toBeNull();

        expect(typeof verifiedToken).toBe('string');
        expect(verifiedToken).toBe(payload);
        expect(typeof verifiedToken).not.toBe('object');
      });
    });

    it('should throw an error while verifying a token with incorrect secret', () => {
      const payload = 'test';
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      jwte.verify(signedToken, 'wrong-secret', encryptionOptions, {}, (err, verifiedToken) => {
        expect(verifiedToken).toBeUndefined();

        expect(err).toBeDefined();
        expect(err).toBeInstanceOf(Error);
        expect(err?.name).toBe('JsonWebTokenError');
        expect(err?.message).toBe('invalid signature');
      });
    });

    it('should throw return void when verifying a token without providing a callback', () => {
      const payload = 'test';
      const jwtSecret = 'secret';

      const encryptionOptions: EncryptionOptions = {
        algorithm: 'aes-256-cbc',
        iv: 'abcd1234abcd1234', // 16-bit string
        key: 'abcd1234abcd1234efgh5678efgh5678', // 32-bit string
      };

      const signedToken = jwte.sign(payload, jwtSecret, encryptionOptions, {
        expiresIn: '1min',
      });

      expect(signedToken).toBeDefined();
      expect(typeof signedToken).toBe('string');

      const response = jwte.verify(signedToken, jwtSecret, encryptionOptions);

      expect(response).toBeUndefined();
    });
  });
});


types/index.ts
import {
  DecodeOptions,
  SignOptions,
  Secret,
  JwtPayload,
  Jwt,
  VerifyOptions,
  VerifyCallback,
  VerifyErrors,
} from 'jsonwebtoken';

export type EncryptionAlgorithm = 'aes-128-cbc' | 'aes-192-cbc' | 'aes-256-cbc';
// TODO: Support the CCM and GCM algorithms
// | 'aes-128-ccm'
// | 'aes-192-ccm'
// | 'aes-256-ccm'
// | 'aes-128-gcm'
// | 'aes-192-gcm'
// | 'aes-256-gcm';

export interface EncryptionOptions {
  /**
   * key - a 16-bit, 24-bit, or 32-bit raw string used by the algorithms. The 16-bit string is used on algorithms that include 128. The 24-bit string is used on algorithms that include 192. The 32-bit string is used on algorithms that include 256
   */
  key: string;
  /**
   * iv - a 16-bit raw string initialization vector
   * Initialization vectors should be unpredictable and unique; ideally, they will be cryptographically random
   * @link https://en.wikipedia.org/wiki/Initialization_vector
   */
  iv: string;
  /**
   * algorithm - The cypher algorithm to be used to  encrypt the payload
   */
  algorithm: EncryptionAlgorithm;
}

export interface DecryptedResult {
  [key: string]: unknown;
}

export interface EncryptedResult {
  data: string;
}

export interface EncryptedData {
  data?: string;
}

export type { DecodeOptions, SignOptions, Secret, JwtPayload, Jwt, VerifyOptions, VerifyCallback, VerifyErrors };
